src/ipc/ipc.c

```diff
@@ -170,6 +170,17 @@ int ipc_comp_free(struct ipc *ipc, uint32_t comp_id)
 
 	/* free component and remove from list */
 	comp_free(icd->cd);
+
+	/* set pipeline sink/source/sched pointers to NULL if needed */
```

Comment:
This violates the rule, where components can be freed in any order.
cd->pipeline can already be NULL.